csrc/scheduler/matmul_utils.cpp

```diff
@@ -1130,6 +1252,25 @@ std::unique_ptr<MatmulParams> getMatmulHeuristics(
         "NVFUSER_MATMUL_HEURISTIC_PLUGIN=/path/to/libmatmulheuristic.so");
   }
 
+  if (mparams->tiling_strategy ==
+      MatmulParams::TilingStrategy::DistributeTilesAcrossSMs) {
+    const int64_t cluster_size =
+        (mparams->cluster_dims.x * mparams->cluster_dims.y *
+         mparams->cluster_dims.z);
+    const int64_t num_sms =
+        at::cuda::getCurrentDeviceProperties()->multiProcessorCount;
+
+    mparams->num_clusters = num_sms / cluster_size;
+
+    if (cluster_size > 2L) {
+      // Large cluster sizes might require lower than 100% occupancy, so we
+      // limit the number of launched CGAs here to guarantee only a single wave
+      // gets launched.
+      mparams->num_clusters = std::min(
```

Comment:
If `num_clusters == -1`, then this is incorrect.

Reply:
This feels like validation. Why can't this go in the scheduler `validate` function?

e.g., We have manual parameters in the tests that won't go through this path.